Applications/Spire/Source/KeyBindings/TaskKeysPage.cpp

```diff
@@ -73,18 +87,155 @@ TaskKeysPage::TaskKeysPage(std::shared_ptr<KeyBindingsModel> key_bindings,
   layout->addWidget(toolbar);
-  layout->addWidget(make_task_keys_table_view(
+  m_table_view = make_task_keys_table_view(
     m_key_bindings->get_order_task_arguments(),
     std::make_shared<LocalComboBoxQueryModel>(), std::move(destinations),
-    std::move(markets)));
+    std::move(markets));
+  m_table_view->installEventFilter(this);
+  layout->addWidget(m_table_view);
   auto box = new Box(body);
   update_style(*box, [] (auto& style) {
     style.get(Any()).set(BackgroundColor(QColor(0xFFFFFF)));
   });
   enclose(*this, *box);
+  m_new_task_form = new LineInputForm(tr("New Task"), *this);
+  m_new_task_form->connect_submit_signal(
+    std::bind_front(&TaskKeysPage::on_new_task_submission, this));
+  update_button_state();
+  auto& row_selection = m_table_view->get_selection()->get_row_selection();
+  m_selection_connection = row_selection->connect_operation_signal(
+    std::bind_front(&TaskKeysPage::on_row_selection, this));
+  m_table_operation_connection =
+    m_table_view->get_body().get_table()->connect_operation_signal(
+      std::bind_front(&TaskKeysPage::on_table_operation, this));
 }
 
 const std::shared_ptr<KeyBindingsModel>&
     TaskKeysPage::get_key_bindings() const {
   return m_key_bindings;
 }
+
+bool TaskKeysPage::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == m_table_view) {
+    if(event->type() == QEvent::KeyPress) {
+      auto& key_event = *static_cast<QKeyEvent*>(event);
+      if(key_event.modifiers() & Qt::ShiftModifier &&
+          (key_event.key() == Qt::Key_Return ||
+            key_event.key() == Qt::Key_Enter)) {
+        on_new_task_action();
+        return true;
+      } else if(key_event.modifiers() & Qt::ControlModifier &&
+          key_event.key() == Qt::Key_D) {
+        on_duplicate_task_action();
+        return true;
+      }
+    }
+  } else if(watched == m_added_region_item) {
+    if(event->type() == QEvent::Show) {
+      auto enter_event =
+        QKeyEvent(QEvent::KeyPress, Qt::Key_Enter, Qt::NoModifier);
+      QCoreApplication::sendEvent(&m_added_region_item->get_body(),
+        &enter_event);
+      m_added_region_item->removeEventFilter(this);
+    }
+  }
+  return QWidget::eventFilter(watched, event);
+}
+
+void TaskKeysPage::update_button_state() {
+  auto is_enabled =
+    m_table_view->get_selection()->get_row_selection()->get_size() > 0;
+  m_duplicate_button->setEnabled(is_enabled);
+  m_delete_button->setEnabled(is_enabled);
+}
+
+void TaskKeysPage::on_new_task_action() {
+  m_new_task_form->show();
+  auto window = m_new_task_form->window();
+  window->move(
+    mapToGlobal(QPoint(0, 0)) + rect().center() - window->rect().center());
+}
+
+void TaskKeysPage::on_duplicate_task_action() {
+  auto& selection = m_table_view->get_selection()->get_row_selection();
+  if(selection->get_size() == 0) {
+    return;
+  }
+  m_table_view->setFocus();
+  auto last_current_row = m_table_view->get_current()->get()->m_row;
+  auto sorted_selection = std::vector<int>();
```

Comment:
Would this work?

    auto sorted_selection = std::vector(selection->begin(), selection->end());